pkg/kubeadm/master/discovery.go

```diff
@@ -0,0 +1,98 @@
+package kubemaster
+
+import (
+	"crypto/x509"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+
+	"k8s.io/kubernetes/pkg/api"
+	unversionedapi "k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	kubeadmapi "k8s.io/kubernetes/pkg/kubeadm/api"
+	"k8s.io/kubernetes/pkg/kubeadm/tlsutil"
+)
+
+type kubeDiscovery struct {
+	Deployment *extensions.Deployment
+	Secret     *api.Secret
+}
+
+func encodeSecretData(params *kubeadmapi.BootstrapParams, caCert *x509.Certificate) map[string][]byte {
+	// TODO ListenIP is probably not the right now, although it's best we have right now
+	// if user provides a DNS name, or anything else, we should use that, may be it's really
+	// the list of all SANs (minus internal DNS names and service IP)?
+
+	var (
+		data         = map[string][]byte{}
+		endpointList = []string{}
+		tokenMap     = map[string]string{}
+	)
+
+	endpointList = append(endpointList, fmt.Sprintf("https://%s:443", params.Discovery.ListenIP))
+	tokenMap[params.Discovery.TokenID] = hex.EncodeToString(params.Discovery.Token)
+
+	data["endpoint-list.json"], _ = json.Marshal(endpointList)
+	data["token-map.json"], _ = json.Marshal(tokenMap)
+	data["ca.pem"] = tlsutil.EncodeCertificatePEM(caCert)
+
+	return data
+}
+
+func createSpec(params *kubeadmapi.BootstrapParams, deploymentName string, secretName string, secretData map[string][]byte) kubeDiscovery {
+	l := map[string]string{"name": deploymentName}
+	kd := kubeDiscovery{
+		Deployment: &extensions.Deployment{
+			ObjectMeta: api.ObjectMeta{Name: deploymentName},
+			Spec: extensions.DeploymentSpec{
+				Replicas: 1,
+				Selector: &unversionedapi.LabelSelector{MatchLabels: l},
+				Template: api.PodTemplateSpec{
+					ObjectMeta: api.ObjectMeta{Labels: l},
+				},
+			},
+		},
+		Secret: &api.Secret{
+			ObjectMeta: api.ObjectMeta{Name: secretName},
+			Type:       api.SecretTypeOpaque,
+		},
+	}
+
+	kd.Deployment.Spec.Template.Spec = api.PodSpec{
```

Comment:
Good question. The assumption was that master is all we have right now, and
nodes won't join without JWS... I am happy to add lables or whatever, to
just make double-sure.

On Wed, 31 Aug 2016, 22:31 Joe Beda, <notifications@github.com> wrote:

> In pkg/kubeadm/master/discovery.go
> <https://github.com/kubernetes/kubernetes/pull/31221#discussion_r77077501>
> :
>
> > +			ObjectMeta: api.ObjectMeta{Name: deploymentName},
> > +			Spec: extensions.DeploymentSpec{
> > +				Replicas: 1,
> > +				Selector: &unversionedapi.LabelSelector{MatchLabels: l},
> > +				Template: api.PodTemplateSpec{
> > +					ObjectMeta: api.ObjectMeta{Labels: l},
> > +				},
> > +			},
> > +		},
> > +		Secret: &api.Secret{
> > +			ObjectMeta: api.ObjectMeta{Name: secretName},
> > +			Type:       api.SecretTypeOpaque,
> > +		},
> > +	}
> > +
> > +	kd.Deployment.Spec.Template.Spec = api.PodSpec{
>
> Shouldn't this pod be pinned to the master node? How else are we going to
> find it?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31221/files/57096ca902c5c666d45df191037d9278972af1fd#r77077501>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS1q7O5O9qzCJLoIguSRajz2RnfoUks5qlfK6gaJpZM4JqsTH>
> .
>